FIX #127: Remove MM_ScanClassesMode and dependencies from OMR

ktbriggs opened the pull request:
New CollectorLanguageInterface methods for ConcurrentGC:
concurrentGC_concurrentInitializationComplete(..)
concurrentGC_kickoffCardCleaning(..)
concurrentGC_startConcurrentScanning(..)
concurrentGC_concurrentScanningStarted(..)
concurrentGC_isConcurrentScanningComplete(..)
concurrentGC_reportConcurrentScanningMode(..)
Provided Java-side implementations for these, stub implementations for
omr example and glue code; OMR_GC_MODRON_CONCURRRENT_MARK enabled
for example/glue but ConcurrentGC not fully integrated with example/glue
yet.

Moved all references to MM_ScanClassesMode (class) and
J9VM_GC_DYNAMIC_CLASS_UNLOADING (build flag) out of OMR source code.

Cleaned up some logic in MM_ConcurrentGC::concurrentMark() (was not
flushing local buffers if interrupted by another thread requesting VM
access).

Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

charliegracie commented on example/glue/configure_includes/configure_common.mk:
```diff
@@ -24,5 +24,6 @@ CONFIGURE_ARGS += \
   --enable-OMR_OMRSIG \
   --enable-fvtest \
   --enable-OMR_GC_SEGREGATED_HEAP \
+  --enable-OMR_GC_MODRON_CONCURRENT_MARK \
```
I would prefer if we disabled this by default until the testing has been updated.  This will also ensure that all new glue layer APIs are properly if defined

ktbriggs replied:
I have basic integration of ConcurrentGC with fvtest -- compiles, runs without fail all current gc tests, instantiating MM_ConcurrentGC instead of MM_ParallelGlobalGC whenever gencon is selected for GC policy. But current test cases are insufficient to drive tests to point of concurrent kickoff and only global GC occurring in gencon test runs is a system GC at end of test. 

We need more work (a model for generating high-volume heaps with mutable object graphs) in fvtest/gctest before they can exercise gencon in significant ways. Meantime Java is consuming current gencon (with MM_ConcurrentGC and MM_Scavenger), and current fvtest/gctest runs fine with concurrentMark enabled. 

Are you sure about this?

charliegracie replied:
I would like to see evidence that new Glue APIs are properly if defined so that if a new language  not using that feature can compile.  Two of our recent features did not do this and you could not compile without those feature flags enabled

charliegracie commented on gc/base/ScanClassesMode.hpp:
```diff
@@ -36,23 +36,57 @@ class MM_ScanClassesMode : public MM_Base
 	volatile uintptr_t _scanClassesMode;
 
 public:
+	typedef enum {
```
Is scanClassesMode something OMR needs to support? This feels like Java code that should be moved into the language

ktbriggs replied:
No, OMR will compile without gc/base/ScanClassesMode.hpp. I left it there for this commit so that changes could be reviewed in place. I intend to move it to Modron in a different task (integrate concurrent GC with fvtest/gctest).

charliegracie replied:
ok.  Are these future commits going to be under the same Issue as this or a different one?  If you plan on using the current Issue then you should change this commit message to remove the fix and just say issue

charliegracie commented on gc/base/standard/ConcurrentGC.cpp:
```diff
@@ -174,19 +174,21 @@ con_helper_thread_proc2(OMRPortLibrary* portLib, void *info)
 	MM_ParallelGlobalGC::reportGCCycleStart(env);
 
-	env->_cycleState = NULL;
+	env->_cycleState = previousCycleState;
```
This is restoring the old cycleState?

ktbriggs replied:
This was an odd fix. Without it, a segfault will occur when MM_ParallelGlobalGC::reportGCStart() references env->_cycleState when it is called from MM_ParallelGlobalGC::internalPreCollect() after calling this method (MM_ConcurrentGC::reportGCCycleStart()). MM_ConcurrentGC also calls reportGCCycleStart() when it kicks off card cleaning -- in that case I reset env->_cycleState to NULL after the call.

charliegracie replied:
This feels like a fix that we do not understand.  I would like to understand how this worked in Java so that we can understand if this is the right fix

ktbriggs replied:
Ping me when you have a few minutes and I will explain why I think this change is necessary (it will be easier to have the source in front of us)

ktbriggs replied:
So after backing out this change to I see this (segfault) again. This particular code path predates my changes and would previously have resulted in segfault whenever MM_ConcurrentGC::internalPreCollect() was called prior to concurrent kickoff. 

Only explanation I can see is that this branch (at ConcurrentGC.cpp line 2944, see gdb trace below) is not normally taken and my changes have created this condition, at least for this test case. So my changes are delaying or preventing concurrent kickoff. Looking into that, but I think this code will need fixing in any case -- this branch will always segfault as it stands.

GC policy for trace below was optavgpause.

```
Breakpoint 17, MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2942
2942            if (executionModeAtGC == CONCURRENT_OFF) {
(gdb) bt
#0  MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2942
#1  0x00007ffff4499ba4 in MM_Collector::preCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at Collector.cpp:202
#2  0x00007ffff449a059 in MM_Collector::garbageCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, callingSubSpace=0x7ffff0107640, allocateDescription=0x7ffff79a0210, gcCode=0, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, context=0x0) at Collector.cpp:301
#3  0x00007ffff46066c6 in MM_MemorySubSpaceFlat::allocationRequestFailed (this=0x7ffff0107640, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, allocationType=MM_MemorySubSpace::ALLOCATION_TYPE_OBJECT, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, previousSubSpace=0x7ffff0107290) at MemorySubSpaceFlat.cpp:143
#4  0x00007ffff4608f92 in MM_MemorySubSpaceGeneric::allocateObject (this=0x7ffff0107290, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, baseSubSpace=0x0, previousSubSpace=0x0, shouldCollectOnFailure=true) at MemorySubSpaceGeneric.cpp:303
#5  0x00007ffff4617a88 in MM_TLHAllocationInterface::allocateObject (this=0x7ffff0136370, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:194
#6  0x00007ffff4617c6a in MM_TLHAllocationInterface::allocateArrayletSpine (this=0x7ffff0136370, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:226
#7  0x00007ffff45cf484 in MM_ArrayletAllocationModel::allocateAndInitializeLayout (this=0x7ffff01365a0, memorySpace=0x7ffff01079b0, env=0x7ffff00426e8, allocDesc=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:122
#8  0x00007ffff45d0fa0 in MM_ArrayletAllocationModel::allocateIndexableObject (this=0x7ffff01365a0, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:315
#9  0x00007ffff42aad65 in J9AllocateIndexableObject (vmThread=0x64b300, clazz=0x689c00, size=1310720, allocateFlags=1) at mgcalloc.cpp:516
#10 0x00007ffff6768994 in VM_BytecodeInterpreter::anewarray (this=0x7ffff79a0500, _sp=@0x7ffff79a0468, _pc=@0x7ffff79a0470) at BytecodeInterpreter.hpp:7283
#11 0x00007ffff6780678 in VM_BytecodeInterpreter::run (this=0x7ffff79a0500, vmThread=0x64b300) at BytecodeInterpreter.hpp:9846
#12 0x00007ffff674c9c1 in bytecodeLoop (currentThread=0x64b300) at BytecodeInterpreter.cpp:90
#13 0x00007ffff67f795e in cInterpreter () at javamisc.s:222
#14 0x0000000000000000 in ?? ()
(gdb) p *_stats
$5 = {
  <MM_Base> = {<No data fields>},
  members of MM_ConcurrentGCStats:
  _executionMode = 1,
  _executionModeAtGC = 1,
  _initWorkRequired = 73728,
  _traceSizeTarget = 1244869,
  _kickoffThreshold = 171628,
  _cardCleaningThreshold = 67610,
  _remainingFree = 0,
  _allocationsTaxed = 0,
  _allocationsTaxedAt0 = 0,
  _allocationsTaxedAt25 = 0,
  _allocationsTaxedAt50 = 0,
  _allocationsTaxedAt75 = 0,
  _allocationsTaxedAt100 = 0,
  _traceSizeCount = 0,
  _cardCleanCount = 0,
  _conHelperTraceSizeCount = 0,
  _conHelperCardCleanCount = 0,
  _completeTracingCount = 0,
  _finalTraceCount = 0,
  _finalCardCleanCount = 0,
  _RSScanTraceCount = 0,
  _RSObjectsFound = 0,
  _threadsScannedCount = 0,
  _threadsToScanCount = 0,
  _concurrentWorkStackOverflowOcurred = false,
  _concurrentWorkStackOverflowCount = 0,
  _completedModes = 0,
  _kickoffReason = NO_KICKOFF_REASON,
  _cardCleaningReason = CARD_CLEANING_REASON_NONE
}
(gdb) p _extensions->scavengerEnabled
$7 = false
(gdb) p _extensions->concurrentMark
$8 = true
(gdb) p _extensions->concurrentKickoffEnabled
$9 = true
(gdb) list 2942,2945
2942            if (executionModeAtGC == CONCURRENT_OFF) {
2943                    /* Not in a concurrent cycle so do perform the super class precollect work */
2944                    MM_ParallelGlobalGC::internalPreCollect(env, subSpace, allocDescription, gcCode);
2945            } else {
(gdb) n
2944                    MM_ParallelGlobalGC::internalPreCollect(env, subSpace, allocDescription, gcCode);
(gdb) s
MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:740
740     {
(gdb) n
741             _cycleState = MM_CycleState();
(gdb) n
742             env->_cycleState = &_cycleState;
(gdb) n
743             env->_cycleState->_gcCode = MM_GCCode(gcCode);
(gdb) n
744             env->_cycleState->_type = _cycleType;
(gdb) n
745             env->_cycleState->_activeSubSpace = subSpace;
(gdb) n
746             env->_cycleState->_collectionStatistics = &_collectionStatistics;
(gdb) n
751             if (!env->_cycleState->_gcCode.isExplicitGC()) {
(gdb) n
752                     if(excessive_gc_normal != _extensions->excessiveGCLevel) {
(gdb) n
758             GC_OMRVMInterface::flushCachesForGC(env);
(gdb) n
759             Assert_MM_true(NULL != env->_cycleState);
(gdb) n
761             _markingScheme->getMarkMap()->setMarkMapValid(false);
(gdb) n
763             if (_extensions->processLargeAllocateStats) {
(gdb) n
768             reportGCCycleStart(env);
(gdb) s
MM_ConcurrentGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ConcurrentGC.cpp:178
178     {
(gdb) n
183             _concurrentCycleState = MM_CycleState();
(gdb) n
184             env->_cycleState = &_concurrentCycleState;
(gdb) n
185             env->_cycleState->_type = _cycleType;
(gdb) n
187             MM_ParallelGlobalGC::reportGCCycleStart(env);
(gdb) s
MM_ParallelGlobalGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1163
1163    {
(gdb) n
1164            OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
(gdb) n
1167            TRIGGER_J9HOOK_MM_OMR_GC_CYCLE_START(
(gdb) n
1174    }
(gdb) list 1167,1173
1167            TRIGGER_J9HOOK_MM_OMR_GC_CYCLE_START(
1168                    _extensions->omrHookInterface,
1169                    env->getOmrVMThread(),
1170                    omrtime_hires_clock(),
1171                    J9HOOK_MM_OMR_GC_CYCLE_START,
1172                    _extensions->getHeap()->initializeCommonGCData(env, &commonData),
1173                    env->_cycleState->_type);
(gdb) n
MM_ConcurrentGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ConcurrentGC.cpp:189
189             env->_cycleState = NULL;
(gdb) n
190     }
(gdb) n
MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:769
769             reportGCStart(env);
(gdb) s
MM_ParallelGlobalGC::reportGCStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1199
1199    {
(gdb) n
1200            uintptr_t scavengerCount = 0;
(gdb) n
1201            OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
(gdb) n
1204            scavengerCount = _extensions->scavengerStats._gcCount;
(gdb) n
1207            Trc_MM_GlobalGCStart(env->getLanguageVMThread(), _extensions->globalGCStats.gcCount);
(gdb) n
1208            Trc_OMRMM_GlobalGCStart(env->getOmrVMThread(), _extensions->globalGCStats.gcCount);
(gdb) n
1210            TRIGGER_J9HOOK_MM_OMR_GLOBAL_GC_START(
(gdb) list 1210,1220
1210            TRIGGER_J9HOOK_MM_OMR_GLOBAL_GC_START(
1211                    _extensions->omrHookInterface,
1212                    env->getOmrVMThread(),
1213                    omrtime_hires_clock(),
1214                    J9HOOK_MM_OMR_GLOBAL_GC_START,
1215                    _extensions->globalGCStats.gcCount,
1216                    scavengerCount,
1217                    env->_cycleState->_gcCode.isExplicitGC() ? 1 : 0,
1218                    env->_cycleState->_gcCode.isAggressiveGC() ? 1: 0,
1219                    _bytesRequested);
1220    }
(gdb) p env->_cycleState
$6 = (MM_CycleState *) 0x0
(gdb) n
$sig = ((siginfo_t*) 0x7ffff799efb0); si_addr = (nil)
(gdb) bt
#0  vmSignalHandler (portLibrary=0x7ffff6f9dec0, gpType=4, gpInfo=0x7ffff799ed90, userData=0x64b300) at gphandle.c:797
#1  0x00007ffff67f22aa in structuredSignalHandler (portLibrary=0x7ffff6f9dec0, gpType=4, gpInfo=0x7ffff799ed90, userData=0x64b300) at gphandle.c:723
#2  0x00007ffff60aabe0 in masterSynchSignalHandler (signal=11, sigInfo=0x7ffff799efb0, contextInfo=0x7ffff799ee80) at ../../omr/port/unix/omrsignal.c:838
#3  <signal handler called>
#4  0x00007ffff449fcc4 in MM_GCCode::isExplicitGC (this=0x0) at GCCode.cpp:32
#5  0x00007ffff455e75e in MM_ParallelGlobalGC::reportGCStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1210
#6  0x00007ffff455ca01 in MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:769
#7  0x00007ffff454bcbf in MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2944
#8  0x00007ffff4499ba4 in MM_Collector::preCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at Collector.cpp:202
#9  0x00007ffff449a059 in MM_Collector::garbageCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, callingSubSpace=0x7ffff0107640, allocateDescription=0x7ffff79a0210, gcCode=0, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, context=0x0) at Collector.cpp:301
#10 0x00007ffff46066c6 in MM_MemorySubSpaceFlat::allocationRequestFailed (this=0x7ffff0107640, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, allocationType=MM_MemorySubSpace::ALLOCATION_TYPE_OBJECT, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, previousSubSpace=0x7ffff0107290) at MemorySubSpaceFlat.cpp:143
#11 0x00007ffff4608f92 in MM_MemorySubSpaceGeneric::allocateObject (this=0x7ffff0107290, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, baseSubSpace=0x0, previousSubSpace=0x0, shouldCollectOnFailure=true) at MemorySubSpaceGeneric.cpp:303
#12 0x00007ffff4617a88 in MM_TLHAllocationInterface::allocateObject (this=0x7ffff0136370, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:194
#13 0x00007ffff4617c6a in MM_TLHAllocationInterface::allocateArrayletSpine (this=0x7ffff0136370, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:226
#14 0x00007ffff45cf484 in MM_ArrayletAllocationModel::allocateAndInitializeLayout (this=0x7ffff01365a0, memorySpace=0x7ffff01079b0, env=0x7ffff00426e8, allocDesc=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:122
#15 0x00007ffff45d0fa0 in MM_ArrayletAllocationModel::allocateIndexableObject (this=0x7ffff01365a0, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:315
#16 0x00007ffff42aad65 in J9AllocateIndexableObject (vmThread=0x64b300, clazz=0x689c00, size=1310720, allocateFlags=1) at mgcalloc.cpp:516
#17 0x00007ffff6768994 in VM_BytecodeInterpreter::anewarray (this=0x7ffff79a0500, _sp=@0x7ffff79a0468, _pc=@0x7ffff79a0470) at BytecodeInterpreter.hpp:7283
#18 0x00007ffff6780678 in VM_BytecodeInterpreter::run (this=0x7ffff79a0500, vmThread=0x64b300) at BytecodeInterpreter.hpp:9846
#19 0x00007ffff674c9c1 in bytecodeLoop (currentThread=0x64b300) at BytecodeInterpreter.cpp:90
#20 0x00007ffff67f795e in cInterpreter () at javamisc.s:222
#21 0x0000000000000000 in ?? ()
(gdb)
```

charliegracie replied:
This code should not be executed prior to concurrent kick off happening.  The concurrent kick off is what should be causing this to happen

charliegracie commented on gc/base/standard/ConcurrentGC.cpp:
```diff
@@ -2207,7 +2196,7 @@ MM_ConcurrentGC::concurrentMark(MM_EnvironmentStandard *env, MM_MemorySubSpace *
 			if(_stats->switchExecutionMode(CONCURRENT_ROOT_TRACING1, CONCURRENT_ROOT_TRACING2)) {
 				/* Signal threads for async callback */
 				bool signaledThreadsForAsyncCallback = false;
-				_cli->concurrentGC_collectRoots(env, CONCURRENT_ROOT_TRACING1, &_scanClassesMode, signaledThreadsForAsyncCallback, taxPaid);
```
This state machine seems very language specific. 

The concurrent collector would know about roots and could inform the language it is time to scan roots and then it would be up to the language to implement a state machine to do different roots if it needs to do them at different time.

ktbriggs replied:
Looking into this. Stayed away from it because concurrent GC owns execution mode and drives state transitions. ROOT_TRACING* states are sequential so I can probably collapse them into a single ROOT_TRACING state where concurrent GC uses an alternate protocol to handshake with client language to manage buffering and helper threads while tracing roots in payAllocationTax().

charliegracie replied:
I would like to see this moved into the language for this commit.

charliegracie commented on gc/include/omrmodroncore.h:
```diff
@@ -101,19 +101,10 @@ enum {
-		NO_LANGUAGE_KICKOFF_REASON=1,
-		FORCED_UNLOADING_CLASSES,
+	NO_LANGUAGE_KICKOFF_REASON=1,
+	FORCED_UNLOADING_CLASSES,
```
The FORCED_UNLOADING_CLASSES should be moved out of OMR and into Java

ktbriggs replied:
Ok.

ktbriggs commented:
@amicic I had similar reservations -- moving CONCURRENT_ROOT_TRACING out of previous order forced a number of changes to complex conditional logic that I have restored by moving CONCURRENT_ROOT_TRACING back to original position and skipping a number (~24) enumerator values that are reserved for client language use. That should be a sufficient range bit it can be extended if required.

charliegracie commented:
@amicic @ktbriggs I agree that the root tracing should continue to happen at the same location and the Enum should be number so that it makes sense.

I do not think we need to reserve states for the language. Concurrent GC asks the language to scan roots and to tell it when it is complete.  If the language has 2, 10, 1000 root scanning states that is up the language.  The language could easily have its own state machine inside the scan roots call.

ktbriggs commented:
@amicic @charliegracie I reserved a range of >1 root tracing substate values for client language for the same reason that Java has >1 root tracing substates -- as I understand it, this is to allow client language to partition its root tracing work into discrete and disjoint work units that can be executed in parallel by >1 mutator threads as they pay allocation tax, without exposing shared _executionMode directly to client language (CLI::concurrentGC_getNextExecutionMode(executionMode) allows client language to determine sequencing of its root tracing states). Am I missing something?

amicic commented:
I'm fine with reserved ranges for the states.

ktbriggs commented:
Most recent commit (50a189d) squashes all work since previous review (ed16fd0) and is ready for review. I have one more commit for this work item (doxygen-like comments for scavenger and concurrent APIs in CollectorLanguageInterface.hpp) that I will submit for subsequent review.

charliegracie commented on example/glue/CollectorLanguageInterfaceImpl.cpp:
```diff
@@ -197,16 +213,12 @@ MM_CollectorLanguageInterfaceImpl::parallelDispatcher_handleMasterThread(OMR_VMT
 void
 MM_CollectorLanguageInterfaceImpl::generationalWriteBarrierStore(OMR_VMThread *omrThread, omrobjectptr_t parentObject, fomrobject_t *parentSlot, omrobjectptr_t childObject)
 {
```
Why is this implementation still in the Impl class?  Is not exactly the same for all languages?

ktbriggs replied:
Will add concurrent/scavenger barriers inline in writeAccessBarrierStore(); unfortunately not possible to inline entire method in CollectorLanguageInterface.hpp as there are circular includes dependencies in that case.

charliegracie commented on example/glue/CollectorLanguageInterfaceImpl.cpp:
```diff
@@ -402,6 +414,13 @@ MM_CollectorLanguageInterfaceImpl::compactScheme_languageMasterSetupForGC(MM_Env
 #endif /* OMR_GC_MODRON_COMPACTION */
 
 #if defined(OMR_GC_MODRON_CONCURRENT_MARK)
+void
```
See previous note on generationalWriteBarrierStore

ktbriggs replied:
See response to previous line note.

charliegracie commented on run_configure.mk:
```diff
@@ -123,7 +123,7 @@ sh configure --disable-auto-build-flag 'OMRGLUE=$(OMRGLUE)' 'SPEC=$(SPEC)' $(CON
 endef
 
-CONFIGURE_DEPENDENCIES := SPEC config.guess config.sub configure toolconfig/configure
+CONFIGURE_DEPENDENCIES := SPEC config.guess config.sub install-sh configure toolconfig/configure
```
The code in this file was removed why are you adding it back?

ktbriggs replied:
This was the result of a merge. I guess I misunderstood Robert's comment (read that he was adding dummy install-sh, instead he removed it). Reverting this to match Robert's commit (a7efc2a).

ktbriggs commented:
Latest commit 527b105 addresses comments (fix up write barriers, undo botched merge for run_configure.mk) and moves some scavenger/concurrentGC-related code from Java into OMR. This should be reviewed with Java changes for RTC 113259 CS2.

If accepted this should complete #127. In that case I will squash this commit onto 50a189d and ed16fd0 and submit for pull to eclipse/master.

charliegracie commented on fvtest/gctest/StartupManagerTestExample.cpp:
```diff
@@ -84,7 +84,12 @@ MM_StartupManagerTestExample::parseLanguageOptions(MM_GCExtensionsBase *extensio
+#if defined(OMR_GC_MODRON_SCAVENGER)
 					extensions->scavengerEnabled = (0 == j9_cmdla_stricmp(attr.value(), "gencon"));
+#if defined(OMR_GC_MODRON_CONCURRENT_MARK)
+					extensions->concurrentMark = extensions->scavengerEnabled;
```
There is no way to enable concurrent global collections without enabling a generational collector?

ktbriggs replied:
Made this change: 
```
extensions->concurrentMark = extensions->scavengerEnabled || (0 == j9_cmdla_stricmp(attr.value(), "optavgpause"));
```
 and added a gctest configuration for this (optavgpause_GC_config.xml). 

All test configurations pass, and ConcurrentGC is instantiated as concurrent collector in optavgpause and gencon configurations, but more work needs to be done to GCTest allocation helper before allocation tax is collected. As mentioned in a previous commit message I intend to get back and finish GCTest integration with ConcurrentGC in a new issue.

ktbriggs commented:
Commit 74f6072 is incremental (enable optavgpause GC poilicy for GCTest) and ready for review. If approved will squash onto 1e7cfa2 and mark as FIX 127.

charliegracie commented:
Everything looks good.
